server/src/main/java/org/elasticsearch/plugins/PluginsUtils.java

```diff
@@ -77,43 +79,111 @@ public static List<Path> findPluginDirs(final Path rootPath) throws IOException
+                            + "] was built for non-semantic Elasticsearch version "
+                            + info.getElasticsearchVersion()
+                            + " and cannot run on version "
+                            + Build.current().version()
+                    );
+                }
+
+                // case: Major version mismatch
+                if (pluginElasticsearchVersion.major != currentElasticsearchVersion.major) {
+                    throw new IllegalArgumentException(
+                        "Stable Plugin ["
+                            + info.getName()
+                            + "] was built for Elasticsearch major version "
+                            + pluginElasticsearchVersion.major
+                            + " but version "
+                            + Build.current().version()
+                            + " is running"
+                    );
+                }
+
+                // case: stable plugin from the future
+                if (pluginElasticsearchVersion.after(currentElasticsearchVersion)) {
+                    throw new IllegalArgumentException(
+                        "Stable Plugin ["
+                            + info.getName()
+                            + "] was built for Elasticsearch version "
+                            + info.getElasticsearchVersion()
+                            + " but earlier version "
+                            + Build.current().version()
+                            + " is running"
+                    );
+                }
+            } else if (info.getElasticsearchVersion().equals(currentElasticsearchVersion.toString()) == false) {
                 throw new IllegalArgumentException(
-                    "Stable Plugin ["
+                    "Plugin ["
                         + info.getName()
                         + "] was built for Elasticsearch version "
                         + info.getElasticsearchVersion()
-                        + " but earlier version "
-                        + Version.CURRENT
+                        + " but version "
+                        + Build.current().version()
                         + " is running"
                 );
             }
-        } else if (info.getElasticsearchVersion().equals(Version.CURRENT) == false) {
-            throw new IllegalArgumentException(
-                "Plugin ["
-                    + info.getName()
-                    + "] was built for Elasticsearch version "
-                    + info.getElasticsearchVersion()
-                    + " but version "
-                    + Version.CURRENT
-                    + " is running"
-            );
         }
         JarHell.checkJavaVersion(info.getName(), info.getJavaVersion());
     }
 
+    private record SemanticVersion(int major, int minor, int bugfix, String suffix) {
+
+        static final Pattern semanticPattern = Pattern.compile("^(\\d+)\\.(\\d+)\\.(\\d+)(\\D?.*)$");
```

Comment:
The trailing portion of this regex can be removed now since Build.current().version() no longer contains qualifier info.